### PR TITLE
Add `stop-redirecting-in-notification-bar` feature

### LINKED ISSUE #2952: Stay on the page after clicking "Done" in the Notifications bar
After opening a notification, I get to the page that triggered it (e.g. a PR or issue I subscribed to). At the top of the page, there is a light blue bar with buttons that act on the notification:

Screenshot_2020-03-29_14-07-14

I'd like to have a quick way of marking the notification as "Done" (and maybe also the other actions) on that page without being redirected back to the Notifications page.

Two possible solutions I can imagine:

1. When clicking on the buttons while pressing the "Alt" keyboard button, the button action is executed, the notification bar disappears, but I stay on the page. When "Alt" is not pressed, the behavior doesn't change.
2. When clicking on the buttons, the button action is executed, the notification buttons change (e.g. after clicking "Done", the only valid action is now "Move to inbox"), but I stay on the page. To go back to notifications or close the notification bar, I have to click the respective button.

Similar to #2546.

### PR DESCRIPTION
**Pull request by FloEdelmann**
Closes #2952. I implemented ~version 2 from the issue since it was easier and more people will discover it~ *a mix of version 1 and 2*.

refined-github-notifications-bar

Test this on this page if you came here from a notification :wink:
Otherwise, go to any other page via your [notifications list](https://github.com/notifications) ([done notifications](https://github.com/notifications?query=is%3Adone) work fine, too).

### CODE REVIEW

**Comment by fregante** [comment]
I don’t want this to be a default. If you’re “Done” with it then why do you need to stay on the page?

As you suggested in the issue, it’s best to do this when alt-clicking it

File: source/features/stop-redirecting-in-notifications-bar.tsx [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+	delegate(document, actionButtonSelector, 'click', onNotificationActionClick);
+}
+
+features.add({
+	id: __filebasename,
+	description: 'Stops redirecting to notification inbox from notification bar actions while holding `Alt`.',
+	screenshot: 'https://user-images.githubusercontent.com/202916/80318782-c38cef80-880c-11ea-9226-72c585f42a51.png'
+}, {
+	include: [
+		pageDetect.isRepo
```
Comment (fregante) [inline source/features/stop-redirecting-in-notifications-bar.tsx:22]: I think
```suggestion
		pageDetect.isPR,
		pageDetect.isIssue
```

File: source/features/stop-redirecting-in-notifications-bar.tsx [hunk 2]
```diff
@@ -0,0 +1,25 @@
+import delegate from 'delegate-it';
+import features from '../libs/features';
+import * as pageDetect from '../libs/page-detect';
+
+function onNotificationActionClick(event: delegate.Event<MouseEvent, HTMLButtonElement>): void {
```
Comment (fregante) [inline source/features/stop-redirecting-in-notifications-bar.tsx:5]: I like that the way this function is written doesn't break successive non-alt clicks. Can you mention this fact in a comment?

File: source/features/stop-redirecting-in-notifications-bar.tsx [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+	const shouldRedirectToInbox = !event.altKey;
+	const actionForm = event.delegateTarget.closest('form')!;
+	actionForm.toggleAttribute('data-redirect-to-inbox-on-submit', shouldRedirectToInbox);
+}
+
+async function init(): Promise<void> {
+	const actionButtonSelector = '.notifications-v2.notification-shelf .js-notification-action button';
```
Comment (fregante) [inline source/features/stop-redirecting-in-notifications-bar.tsx:12]: I think only this is needed:

```suggestion
	const actionButtonSelector = '.notification-shelf .js-notification-action button';
```

File: source/features/stop-redirecting-in-notifications-bar.tsx (hunk 1 above)
Comment (FloEdelmann) [reply source/features/stop-redirecting-in-notifications-bar.tsx:22]: At least commits and releases can be notification targets, too.

Comment (fregante) [reply source/features/stop-redirecting-in-notifications-bar.tsx:24]: Indeed. I then we need a new detection function in the current file. Maybe:

```js
const hasNotification = () => location.search.startsWith('?notification_referrer_id=');
```

And then:

```js
{
	include: [
		hasNotification
	],
	waitForDomReady: false,
	repeatOnAjax: false,
	init
}
```

Comment (FloEdelmann) [reply source/features/stop-redirecting-in-notifications-bar.tsx:24]: This won't work if you reload on a page with notification bar, as the URL query parameter is removed.

E.g. `https://github.com/sindresorhus/refined-github/pull/3036?notification_referrer_id=XXXXXXXX#pullrequestreview-401122761` will be detected. That turns to `https://github.com/sindresorhus/refined-github/pull/3036#pullrequestreview-401122761` immediately, so after a reload, that page will not be detected, but the notifications bar is still there.

Comment (fregante) [reply source/features/stop-redirecting-in-notifications-bar.tsx:24]: > so after a reload [...] the notifications bar is still there.

That doesn't match my experience:

demo2

From what I see, GitHub uses exactly the same `location.search` test

Comment (FloEdelmann) [reply source/features/stop-redirecting-in-notifications-bar.tsx:24]: Weird, when I reload this page, the notifications bar re-appears after a short delay. I also tried disabling all GitHub-related browser extensions (Refined GitHub, OctoLinker and Notifications Preview for GitHub), but my experience doesn't change.

Comment (fregante) [reply source/features/stop-redirecting-in-notifications-bar.tsx:24]: > It apparently only re-appears on PRs, not issues or releases.

Indeed. Weirdly enough, PRs store this state in `sessionStorage`. Maybe:

```js
const hasNotificationBar = () =>
	location.search.startsWith('?notification_referrer_id=') ||
	JSON.parse(sessionStorage.notification_shelf ?? '{}').pathname === location.pathname;
```

Comment (FloEdelmann) [reply source/features/stop-redirecting-in-notifications-bar.tsx]: That works fine!

File: source/libs/page-detect.ts [hunk 4]
```diff
@@ -414,3 +414,8 @@ export const _isActionPage = [
 	'https://github.com/marketplace/actions/github-action-for-assignee-to-reviewer',
 	'https://github.com/marketplace/actions/hugo-actions'
 ];
+
+export const _hasNotificationBar = skip;
+export const hasNotificationBar = (): boolean =>
+	location.search.startsWith('?notification_referrer_id=') ||
+	JSON.parse(sessionStorage.notification_shelf ?? '{}').pathname === location.pathname;
```
Comment (fregante) [inline source/libs/page-detect.ts:421]: This is too specific to be part of `page-detect`, can you move it to the feature’s file?

Comment (FloEdelmann) [reply source/libs/page-detect.ts]: Done.

**Comment by fregante** [comment]
Thanks Flo!

**Comment by FloEdelmann** [comment]
@fregante Thanks for reviewing and merging :blush:

**Comment by tonglil** [comment]
My reason for wanting it to stop redirecting is that I often open many tabs at once, either through 
image
or ctrl-clicking the ones I need to check "as a batch".

The redirection is weird because I don't know how long to wait before it has marked it as done so it is safe to close the tab, but the redirect is slow.

Ideal workflow: press "done" + ctrl+w to move on to the next tab
Not ideal: press "done", wait for redirect to start, ctrl+w -> this just feels slow and jarring (two browser paints in quick succession)

**Comment by karlhorky** [comment]
Thanks for this feature!

I would like to voice my support for having an option to turn this on **without** holding `alt`. I do agree with @fregante that it should be an option, not turned on by default.

**Comment by fregante** [comment]
1. This extension only tries to improve the experience, but this seems just _different_ than the expected/standard behavior.
2. We don't have ["off by default" features](https://github.com/sindresorhus/refined-github/issues/2960#issuecomment-606765382)

Let's discuss on #3173

**Pull request merged**